pex/environment.py

```diff
@@ -235,20 +242,28 @@ def activate(self):
     return self._working_set
 
   def _resolve(self, working_set, reqs):
-    reqs_by_key = OrderedDict((req.key, req) for req in reqs)
-    unresolved_reqs = OrderedDict()
-    resolveds = OrderedSet()
-
     environment = self._target_interpreter_env.copy()
     environment['extra'] = list(set(itertools.chain(*(req.extras for req in reqs))))
 
-    # Resolve them one at a time so that we can figure out which ones we need to elide should
-    # there be an interpreter incompatibility.
-    for req in reqs_by_key.values():
+    reqs_by_key = OrderedDict()
+    for req in reqs:
       if req.marker and not req.marker.evaluate(environment=environment):
         TRACER.log('Skipping activation of `%s` due to environment marker de-selection' % req)
         continue
-      with TRACER.timed('Resolving %s' % req, V=2):
+      reqs_by_key.setdefault(req.key, []).append(req)
+
+    unresolved_reqs = OrderedDict()
+    resolveds = OrderedSet()
+
+    # Resolve them one at a time so that we can figure out which ones we need to elide should
+    # there be an interpreter incompatibility.
+    for key, reqs in reqs_by_key.items():
+      with TRACER.timed('Resolving {} from {}'.format(key, reqs), V=2):
+        # N.B.: We resolve the bare requirement with no version specifiers since the resolve process
+        # used to build this pex already did so. There may be multiple distributions satisfying any
+        # particular key (e.g.: a Python 2 specific version and a Python 3 specific version for a
+        # multi-python PEX) and we want the working set to pick the most appropriate one.
+        req = Requirement.parse(key)
```

Comment:
Yes, this is a build time / run time distinction.

These are requirements stored in a built PEX's metadata, and all of those requirements are of the form `[key]==[version](; [markers])'. They are this way since they're buit from the _results_ of a resolve. That resolve does accept the myriad forms of requirement strings, but it plops out concrete dists. Those concrete dists are used to create the requirements in the PEX metadata used here to boot up at PEX runtime:
https://github.com/pantsbuild/pex/blob/a07954a1bce1cf0c023c67f36e4e37f078eb23ba/pex/resolver.py#L555-L579

Which uses:
https://github.com/pantsbuild/pex/blob/a07954a1bce1cf0c023c67f36e4e37f078eb23ba/pex/resolver.py#L96-L97

Which ues:
https://github.com/pantsbuild/pex/blob/a07954a1bce1cf0c023c67f36e4e37f078eb23ba/pex/vendor/_vendored/setuptools/pkg_resources/__init__.py#L2870-L2877